README.rst
=====
holoviews_functions
=====

Functions that produce standardized plots

* Free software: MIT license

Prerequisites:

* pyviz projects holoviews and panel


Brief Description:

* statsplots.py  - horizontal, vertical and oblique line segments, stem plots, regression plots, forest plots
* timefreq.py    - time series plots, interpolated scalograms with contours
* ngons.py       - nGons and nGonal boundaries (uses hv.Ellipse - I'd like to figure out how to package these as an hv Element)


HISTORY.rst
=======
History
=======

0.1 (2021-5-5)
------------------

* initial set of functions


holoviews_functions/ngons.py
import numpy as np

import holoviews as hv; hv.extension('bokeh', logo=False)
import panel as pn; pn.extension()

# ============================================================================
def ngon_data( x=0,y=0, specs=[(1,1)], orientation=0., arc=(0,2*np.pi), npoints=3 ):
    ''' compute coordinate locations for n-gons with vertices on an ellipse

    Args:
        x           : x coordinate of n-gon center
        y           : y coordinate of n-gon center
        specs       : array of (a,b) lengths of principal axes of the ellipse(s)
        orientation : inclination of the principal axis of the ellipse in radians
        arc         : (a1,a2) angles in radians of the subtended arc of the n-gon vertices
        npoints     : number of vertices

    Notes:
        | since the resulting paths are closed, the computation yields npoints+1 vertices;
        | thus, arcs that do not correspond to a full circle result in an extra vertex
    '''
    t_vals = np.linspace(*arc,npoints+1)
    s = np.sin(t_vals)
    c = np.cos(t_vals)
    rot  = np.array([[np.cos(orientation), -np.sin(orientation)],
                     [np.sin(orientation),  np.cos(orientation)]])
    gons = []
    for (a,b) in specs:
        data = np.array(list(zip(a*c,b*s)))
        gons.append( np.tensordot(rot, data.T, axes=[1,0]).T+np.array([x,y]))

    return gons
# ----------------------------------------------------------------------------
def ngon_ring(x=0, y=0, specs=[(1,1),(.9,.9)], arc=(0,2*np.pi), orientation=0, npoints=40,
              line_color='k', line_alpha=1, fill_color='k', fill_alpha=0., label=''):
    ''' an ngon outline with finite width
    '''
    l = ngon_data(x,y, specs=specs, arc=arc,orientation=orientation, npoints=npoints)
    d = ( np.hstack([ l[0][:,0],l[1][::-1,0]] ), np.hstack([ l[0][:,1],l[1][::-1,1]] ) )
    
    o_b = {'Polygons' : dict( color=fill_color, alpha=fill_alpha, line_alpha=0), 'Path': dict(line_color=line_color, line_alpha=line_alpha) }
    o_m = {'Polygons' : dict( color=fill_color, alpha=fill_alpha),               'Path': dict(color=line_color,      alpha=line_alpha) }
    return (hv.Polygons([d], label=label)*hv.Path([l[0], l[1]])).opts( o_b, backend='bokeh' ) #.opts( o_m, backend='matplotlib ')
#    return hv.Polygons([d], label=label).opts( color=fill_color, alpha=fill_alpha, line_alpha=0 )*\
#           hv.Path([l[0], l[1]]).opts( line_color=line_color, line_alpha=line_alpha)

# ----------------------------------------------------------------------------
def ngon(x=0,y=0, spec=(1,1), orientation=0., arc=(0,2*np.pi), npoints=40,\
         line_color='k', line_alpha=1., color='k', alpha=0., label=''):
    ''' an ngon '''
    e = ngon_data(x,y,specs=[spec], arc=arc,orientation=orientation,npoints=npoints)

    return hv.Polygons(e, label=label).opts( line_color=line_color,line_alpha=line_alpha, color=color, alpha=alpha)


holoviews_functions/statsplots.py
import holoviews as hv
from holoviews.operation.datashader import aggregate, shade, datashade, dynspread
import numpy as np
import pandas as pd

# ===========================================================================
import unicodedata as ucd
class DimName(object):
    '''functions to facilitate accessing unicode letters'''
    # https://en.wikipedia.org/wiki/Unicode_subscripts_and_superscripts
    @staticmethod
    def lcg(s, n=None):
        '''lower case greek letters with subscript lcg('alpha',1)'''
        l = ucd.lookup('greek small letter ' + s) 
        if n is None: return l
        else:         return '{0}{1}'.format(l,chr(0x2080 + n))
    @staticmethod
    def ucg(s, n=None):
        '''upper case greek letters with subscript ucg('sigma',1)'''
        l = ucd.lookup('greek capital letter ' + s) 
        if n is None: return l
        else:         return '{0}{1}'.format(l,chr(0x2080 + n))
    @staticmethod
    def subscript(s,n): return '{0}{1}'.format(s,chr(0x2080 + n))

# ===========================================================================
def forest_plot( df = dict( mean    = np.array( [2.0, 4.0, 6.0,   8.0 ]),
                            sd      = np.array( [0.5, 0.5, 0.5,   0.5 ]),
                            low     = np.array( [0.0, 2.0, 4.0,   6.0 ]),
                            hgh     = np.array( [3.0, 5.0, 7.0,   9.0 ]) ),
                 support =.95,
                 index   = [DimName.lcg('alpha'), DimName.lcg('beta', 0),DimName.lcg('beta', 1),DimName.lcg('sigma')],
                 offset  = 0., group = None):
    rng   = (np.floor(df['low'].min()), np.ceil(df['hgh'].max()) )  # improve: this does not work if all data << 1
    d_rng = (rng[1]-rng[0])*.01
    rng   = (rng[0]-d_rng, rng[1]+d_rng)

    e_u = df['hgh' ] - df[ 'mean']
    e_l = df['mean'] - df[ 'low' ]
    
    ticks = [ (i,s) for i,s in enumerate(index) ]
    vals  = np.arange(len(df['mean']))+offset

    specs = {
        'ErrorBars': dict(invert_axes=True, shared_axes=False, show_grid=True,  width=600, height=200,
                          color='slateblue',axiswise=True),
        'Points':    dict(invert_axes=True, tools=['hover'], toolbar='above', shared_axes=False,
                          color='slateblue', marker='+', size=10, line_width=3,
                          axiswise=True),
        'HLine':     dict(color='coral', alpha=0.4, line_width=1.2)
    }
    if group is not None: g={'group':group}
    else:                 g={}

    # Use invisible (i.e., alpha=0) markers to enable hover information
    h = \
        hv.ErrorBars( [i for i in zip(vals,  df['mean' ], e_l, e_u)],
                      kdims=['parameter'], vdims=['value','e_l','e_u'], **g)\
          .opts(line_width=1, yticks=ticks ) \
      * hv.ErrorBars( (vals,  df['mean' ], df['sd' ] ),
                      kdims=['parameter'], vdims=['value','sd' ], **g)\
          .opts(line_width=5, lower_head=None, upper_head=None) \
      * hv.Points( (vals,  df['mean']), ['parameter', 'value'], **g ) \
      * hv.Points( (vals,  df['low' ]), ['parameter', 'value'], **g ).opts(alpha=0) \
      * hv.Points( (vals,  df['hgh' ]), ['parameter', 'value'], **g ).opts(alpha=0) \
      * hv.HLine(0, **g)

    return h.opts(specs).redim.range(parameter=(-.5,len(df['mean'])-.5),value=rng).relabel('%d%% Credible Intervals'%int(100.*support))
# ===========================================================================
def vertical_lines(x,y1,y2):
    vlines = [ np.array( [[x[i], y1[i]], [x[i], y2[i]]]) for i in range(len(y1)) ]
    return hv.Path( vlines ).opts( show_legend=True, muted_alpha=0.)
# ---------------------------------------------------------------------------
def horizontal_lines( x1,x2,y):
    hlines = [ np.array( [[x1[i], y[i]], [x2[i], y[i]]]) for i in range(len(x1)) ]
    return hv.Path( hlines ).opts( show_legend=True, muted_alpha=0.)
# ---------------------------------------------------------------------------
def oblique_lines( x1,y1, x2,y2):
    hlines = [ np.array( [[x1[i], y1[i]], [x2[i], y2[i]]]) for i in range(len(x1)) ]
    return hv.Path( hlines ).opts( show_legend=True, muted_alpha=0.)
# ---------------------------------------------------------------------------
def stem(data, curve=True, marker=True):
    '''stem plot:   data=(x,y,e)'''
    x,y,e=data
    vlines = [ np.array( [[x[i], y[i]], [x[i], e[i]]]) for i in range(len(x)) ]

    hs = hv.Path( vlines ).opts( show_legend=True, muted_alpha=0.)

    if marker: hs = hs * hv.Scatter((x,e)).opts(size=4)
    if curve:  hs = hs * hv.Curve((x,y)).opts(line_width=0.8)
    return hs
# ===========================================================================
def jittered_points(data, jitter=0.05, shade=True, shader_opts=dict(cmap="blues")):
    '''given (x_vals[N], y_meas[NumExperiments, N], apply a jitter to the x values
       and possible datashade the resulting scatter plot'''
    
    from holoviews.operation.datashader import datashade
    x_vals, y_meas    = data
    NumExperiments,N  = y_meas.shape

    x_jitter  = np.random.uniform(low=-jitter, high=jitter, size=(N*NumExperiments))
    h_scatter = hv.Scatter( ( np.repeat(x_vals, NumExperiments)+x_jitter, y_meas.T.flatten() ))
    if shade: h_scatter = datashade(h_scatter, **shader_opts)
    return h_scatter
# ---------------------------------------------------------------------------
def regression_lines_spread(data, meas_stats, estim_stats):
    '''given (x_vals[N], y_vals[NumExperiments, N], return hv.Spread overlaid with jittered values'''

    x_vals, y_meas           = data
    y_meas_ave,  y_meas_std  = meas_stats
    y_estim_ave, y_estim_std = estim_stats

    NumExperiments,N  = y_meas.shape

    h_spread_pts  = hv.Spread((x_vals, y_estim_ave, y_estim_std))*\
                    hv.Spread((x_vals, y_meas_ave,  y_meas_std))*\
                    hv.Curve( (x_vals, y_estim_ave))
    return h_spread_pts



holoviews_functions/vectorfield.py
import numpy as np
import holoviews as hv

def vector_field_with_background(x,y, slopes, length=1, subsample=3):
    """
    Given vectors x, y and an array of slopes(x,y)
    return a grayscale image of the slopes y' together with
           a vector field representation of the slopes
    """

    # compute the data
    theta = np.arctan2( slopes, 1)       # angle of the slope y' at x,y

    # Obtain the vector field (subsample the grid)
    decim_x       = x[::subsample]
    decim_y       = y[::subsample]
    decim_theta   = theta[::subsample,::subsample]
    
    if isinstance( length, (int,float)):
        decim_len  = np.full_like( decim_theta, length)
    else:
        decim_len  = length[::subsample, ::subsample]
    vf_opts       = dict(size_index=3, alpha=0.3, muted_alpha=0.05)
    vec_field     = hv.VectorField((decim_x,decim_y,decim_theta,decim_len) ).opts(**vf_opts)

    # Normalize the given array so that it can be used with the RGB element's alpha channel   
    def norm(arr):
        arr = (arr-arr.min())
        return arr/arr.max()

    normXY    = norm(np.copy(slopes))
    img_field = hv.RGB( (x, y, normXY, normXY, normXY, np.full_like(theta, 0.1)),
                        vdims=['R','G','B','A'] )\
                .opts(shared_axes=False)

    return img_field*vec_field



holoviews_functions/timefreq.py
import numpy as np

import holoviews as hv; hv.extension('bokeh', logo=False)
import panel as pn; pn.extension()

# ----------------------------------------------------------------------
def set_toolbar_autohide(plot, element):
    '''autohide the bokeh toolbar. Use with hv.opts(  hooks=[set_toolbar_autohide]'''
    bokeh_plot = plot.state
    bokeh_plot.toolbar.autohide = True

# ======================================================================
def spikes(data, y_base=0, dims=["Time", "x"], label="Signal", curve=True):
    '''plot stems and a curve'''
    if isinstance(data, tuple):
        t,s=data
    else:
        t=np.arange(0,len(data), 1)
        s=data

    vlines = [ np.array( [[t[i], y_base], [t[i], s[i]]]) for i in range(len(s)) ]

    hs = hv.Path( vlines, kdims=dims, label=label ).opts( show_legend=True, muted_alpha=0., color='black')

    if curve: hs = hs * hv.Curve((t,s), dims[0], dims[1], label=label).opts(line_width=0.8)
    return hs

def cx_spikes(data, dims=["Time", "x"], y_base=0, curve=False):
    '''plot stems and a curve for complex data'''
    if isinstance(data, tuple):
        t,s=data
    else:
        t=np.arange(0,len(data), 1)
        s=data
    h = hv.Overlay([
            *spikes((t, np.real(s)), y_base=y_base, label='real', curve=True).opts( "Path", color='blue'),
            *spikes((t, np.imag(s)), y_base=y_base, label='imag', curve=True).opts( "Path", color='red')])
    return h

# ========================================================================
def interpolated_scalogram( x,y, d, method='nearest'):
    ''' interpolate the values of a scalogram'''
    from scipy.interpolate import griddata

    x_lin = np.linspace( x[np.argmin(x)], x[np.argmax(x)], len(x), endpoint=True)

    ndx_ymin = np.argmin(y)
    ndx_ymax = np.argmax(y)
    if ndx_ymax == 0:
        y_lin = np.linspace( y[ndx_ymin], y[ndx_ymax] + 0.25*(y[0]-y[1]), len(y), endpoint=True)
    else:
        y_lin = np.linspace( y[ndx_ymin], y[ndx_ymax] + 0.25*(y[ndx_ymax]-y[ndx_ymax-1]), len(y), endpoint=True)

    xx,yy = np.meshgrid(x, y)
    xi,yi = np.meshgrid(x_lin,y_lin)

    di = griddata( (xx.ravel(), yy.ravel()), d.ravel(), (xi.ravel(), yi.ravel()), method=method )
    return xi,yi, di.reshape(d.shape)

def estimate_levels( d, l = np.sin(0.5*np.pi*np.linspace( 0.1, 1, 5, endpoint=False)), decimals=2 ):
    '''obtain a set of labels for a contour plot'''
    dmin,dmax = np.min(d), np.max(d)
    return np.round( [dmin + t*(dmax-dmin) for t in l], decimals)

def plot_qmesh_contours(t, freqs, tfr, l = None, filled=True, overlaid=True, img_args={}, contour_args={} ):
    '''plot an interpolated QuadMesh as an Image overlaid with contours'''
    from holoviews.operation import contours
    ti,freqsi,tfri = interpolated_scalogram( t, freqs, tfr, 'cubic')
    img = hv.Image( (ti[0,:], freqsi[:,0], tfri), ["Time", "Frequency"] )\
            .opts(cmap="Viridis", width=400,show_grid=False)\
            .opts( **img_args )
    if l is None: l = estimate_levels(tfr)
    return contours(img, levels=l, filled=filled, overlaid=overlaid).opts("Contours", **contour_args)
